File: julia/Euler/src/bin/p0031.jl

# project euler: problem 31

module Prob0031

function solve_0031(coins::Vector{Int} = [1, 2, 5, 10, 20, 50, 100, 200], target::Int = 200)
    tbl = zeros(Int, target + 1)
    tbl[1] = 1

    for c in coins
        for i in c:target
            tbl[i + 1] += tbl[i - c + 1]
        end
    end
    tbl[end]
end

end #module

using .Prob0031: solve_0031
export solve_0031

File: clojure/src/project_euler/sol/p0031.clj
(ns project-euler.sol.p0031)

(defn solve
  ([]
   (solve [1 2 5 10 20 50 100 200] 200))
  ([coins target]
   (let [tbl (long-array (inc target) 0)]
     (aset tbl 0 1)
     (doseq [c coins
             i (range c (inc target))]
       (aset tbl i (+ (aget tbl i) (aget tbl (- i c)))))
     (aget tbl target))))
